micropython/utop/utop.py

```diff
@@ -0,0 +1,108 @@
+                tasks.sort(key=lambda t: t[0])
+                for (
+                    task_id,
+                    task_name,
+                    task_state,
+                    task_priority,
+                    task_runtime,
+                    task_stackhighwatermark,
+                    task_coreid,
+                ) in tasks:
+                    task_runtime_percentage = "-"
+                    if previous_total_runtime != 0 and task_id in previous_task_runtimes:
+                        task_runtime_percentage = "{:.2f}%".format(
+                            100
+                            * time.ticks_diff(task_runtime, previous_task_runtimes[task_id])
+                            / time.ticks_diff(total_runtime, previous_total_runtime)
+                        )
+                    task_state_name = "unknown"
+                    if task_state >= 0 and task_state < len(esp32_task_state_names):
+                        task_state_name = esp32_task_state_names[task_state]
+
+                    print(
+                        "{:>7}  {:>4d}  {:>8d}  {:<9}  {:<14d}  {}{}\x1b[K".format(
+                            task_runtime_percentage,
+                            task_coreid,
+                            task_priority,
+                            task_state_name,
+                            task_stackhighwatermark,
+                            task_name if task_id not in thread_names else thread_names[task_id],
+                            " (*)" if task_id == current_thread_id else "",
+                        )
+                    )
+                    line_count += 1
+
+                    previous_task_runtimes[task_id] = task_runtime
+        else:
+            print("INFO: Platform does not support listing active tasks.\x1b[K")
+            line_count += 1
+
+        print("\x1b[K")
+        line_count += 1
+        print("MicroPython ", end="")
+        micropython.mem_info()
+        line_count += 3
+
+        if esp32 is not None:
+            print("\x1b[K")
+            line_count += 1
+            for name, cap in (("data", esp32.HEAP_DATA), ("exec", esp32.HEAP_EXEC)):
+                heaps = esp32.idf_heap_info(cap)
+                print(
+                    "IDF heap ({}): {} regions, {} total, {} free, {} largest contiguous, {} min free watermark\x1b[K".format(
+                        name,
+                        len(heaps),
+                        sum((h[0] for h in heaps)),
+                        sum((h[1] for h in heaps)),
+                        max((h[2] for h in heaps)),
+                        sum((h[3] for h in heaps)),
+                    )
+                )
+                line_count += 1
+
+        if previous_line_count > line_count:
+            for _ in range(previous_line_count - line_count):
+                print("\x1b[K")
+            print("\x1b[{}A".format(previous_line_count - line_count), end="")
+
+        previous_total_runtime = total_runtime
```

Comment:
On non-esp32 boards, or esp32 without `idf_task_stats()`, `total_runtime` is undefined.